src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/location/importer/LocationsImporter.kt

```diff
@@ -20,33 +17,18 @@ class LocationsImporter(private val repo: LocationRepository, private val eventP
     @Value("\${import-files.locations}")
     private lateinit var locationsFile: String
 
-    fun import(workbook: XSSFWorkbook) : Set<String> {
+    fun import(locationsWorkbook: XSSFWorkbook) : Set<String> {
 
         val locationsFromCells = listOf(Court, Police, Prison, Hospital, Immigration, STCSCH, Other)
         val errors: MutableSet<String?> = mutableSetOf()
+
         locationsFromCells.forEach { locationFromCells ->
-            val sheet = workbook.getSheetAt(locationFromCells.sheetIndex)
-            val rows = sheet.iterator().asSequence().toList().drop(1).filter { r ->
-                !r.getCell(1)?.stringCellValue.isNullOrBlank()
-            }
-            rows.forEach { r ->
-                val rowCells = r.iterator().asSequence().toList()
-                val locationResult = locationFromCells.getLocationResult(rowCells)
-
-                locationResult.fold({ location ->
-                    try {
-                        repo.save(location)
-                    } catch (e: Exception) {
-                        when(e) {
-                            is DataIntegrityViolationException -> {
-                                // TODO collect these errors
-                            }
-                            else -> errors.add(e.message+" "+e)
-                        }
-                    }
-                }, { error -> errors.add(error.message + " " + error.cause.toString())
-                })
+            val sheet = locationFromCells.sheet(locationsWorkbook)
+            val rows = sheet.drop(1).filter { !it.getCell(1)?.stringCellValue.isNullOrBlank() }
 
+            rows.forEach {
+                Result.runCatching { repo.save(locationFromCells.location(it.toList())) }
+                        .onFailure { errors.add(it.message + " " + it.cause.toString()) }
```

Comment:
Nice - could change that `locationCells.forEach` into a map as it's just performing transformations